core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -106,15 +106,13 @@ class HadoopMapReduceCommitProtocol(
     val filename = getFilename(taskContext, ext)
 
     val stagingDir: Path = committer match {
-      case _ if dynamicPartitionOverwrite =>
-        assert(dir.isDefined,
-          "The dataset to be written must be partitioned when dynamicPartitionOverwrite is true.")
-        partitionPaths += dir.get
-        this.stagingDir
       // For FileOutputCommitter it has its own staging path called "work path".
       case f: FileOutputCommitter =>
+        handleDynamicPartitionOverwrite(dir)
         new Path(Option(f.getWorkPath).map(_.toString).getOrElse(path))
-      case _ => new Path(path)
+      case _ =>
+        handleDynamicPartitionOverwrite(dir)
```

Comment:
If both case branches need to call `handleDynamicPartitionOverwrite`, then we can call it outside case match?

Reply:
Since we leverage FileOutputCommitter to commit task attempt file, we only need to handle this in 'case f: FileOutputCommitter' branch

Reply:
OK

Reply:
Since this pr leverages FileOutputCommitter to deal with commit collision, only handle dynamic partition overwrite in the first branch.
BTW, it seems all committers practically used in Spark are derived from 'FIleOutputCommiter'